update gdbm to work with the newer version of gdbm-sys

miquels opened the pull request:
After merging this, you'll have to fix up https://github.com/cholcombe973/gdbm/commit/42976c635c637764385788ab6a00100fbd47b5e9 . to not point at my github repo of course.

miquels committed:
update gdbm for a newer version of gdbm-sys

Access gdbm_errno using *gdbm_errno_location(), which is a thread-local
variable, instead of the global gdbm_errno.

Because of that, we can now mark Gdbm as Send.

miquels committed:
update bitflags to 1.x

miquels committed:
update for modern Error trait

miquels committed:
fixes for recent bitflags crate

The latest `bitflags` does not export the flags at the root crate
namespace anymore. So Gdbm::READER now has to be Gdbm::Open::READER.

miquels committed:
use git repo location for gdbm-sys dependency

cholcombe973 commented:
Thanks for the contribution! I think this looks good. Have you tried running valgrind over it? C bindings always cause me to introduce subtle bugs and I can't tell if this is completely correct just by looking at it.

miquels commented:
No, but I can do that. You mean running it over the test in test/test.rs, or do youhave something more substantial in mind?

cholcombe973 commented:
Yeah i think running it over test.rs would be fine. If it's clean it should be good to go :)

miquels commented:
I ran the test under valgrind and it did not give any errors. Well, two warnings about `statx`, but I happened to look at the stat/statx implementation in the standard library, and that's simply the stdlib probing to see if there's statx support. It uses a NULL buffer for that, then checks if it get ENOSYS (system call not supported) or "anything else" (system call supported).

```
$ ls -l target/debug/deps/test-fd9873c3349180f5
-rwxrwxr-x 1 mikevs mikevs 5631608 Jan 20 20:43 target/debug/deps/test-fd9873c3349180f5*
$ valgrind target/debug/deps/test-fd9873c3349180f5
==751552== Memcheck, a memory error detector
==751552== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==751552== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==751552== Command: target/debug/deps/test-fd9873c3349180f5
==751552==
==751552== Syscall param statx(file_name) points to unaddressable byte(s)
==751552==    at 0x49B589D: syscall (syscall.S:38)
==751552==    by 0x1A227B: statx (weak.rs:95)
==751552==    by 0x1A227B: std::sys::unix::fs::try_statx (fs.rs:123)
==751552==    by 0x19D517: std::sys::unix::fs::stat (fs.rs:1022)
==751552==    by 0x16B9EC: metadata<&std::path::PathBuf> (fs.rs:1567)
==751552==    by 0x16B9EC: term::terminfo::searcher::get_dbpath_for_term (searcher.rs:50)
==751552==    by 0x16B226: term::terminfo::TermInfo::from_name (mod.rs:83)
==751552==    by 0x16B02B: term::terminfo::TermInfo::from_env (mod.rs:69)
==751552==    by 0x1744D4: new<std::io::stdio::Stdout> (mod.rs:219)
==751552==    by 0x1744D4: term::stdout (lib.rs:62)
==751552==    by 0x14F7F9: test::console::run_tests_console (console.rs:252)
==751552==    by 0x15D976: test::test_main (lib.rs:121)
==751552==    by 0x15ED18: test::test_main_static (lib.rs:140)
==751552==    by 0x13CCD5: test::main (in /home/mikevs/devel/rust/gdbm/target/debug/deps/test-fd9873c3349180f5)
==751552==    by 0x13CF3A: core::ops::function::FnOnce::call_once (function.rs:227)
==751552==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==751552==
==751552== Syscall param statx(buf) points to unaddressable byte(s)
==751552==    at 0x49B589D: syscall (syscall.S:38)
==751552==    by 0x1A227B: statx (weak.rs:95)
==751552==    by 0x1A227B: std::sys::unix::fs::try_statx (fs.rs:123)
==751552==    by 0x19D517: std::sys::unix::fs::stat (fs.rs:1022)
==751552==    by 0x16B9EC: metadata<&std::path::PathBuf> (fs.rs:1567)
==751552==    by 0x16B9EC: term::terminfo::searcher::get_dbpath_for_term (searcher.rs:50)
==751552==    by 0x16B226: term::terminfo::TermInfo::from_name (mod.rs:83)
==751552==    by 0x16B02B: term::terminfo::TermInfo::from_env (mod.rs:69)
==751552==    by 0x1744D4: new<std::io::stdio::Stdout> (mod.rs:219)
==751552==    by 0x1744D4: term::stdout (lib.rs:62)
==751552==    by 0x14F7F9: test::console::run_tests_console (console.rs:252)
==751552==    by 0x15D976: test::test_main (lib.rs:121)
==751552==    by 0x15ED18: test::test_main_static (lib.rs:140)
==751552==    by 0x13CCD5: test::main (in /home/mikevs/devel/rust/gdbm/target/debug/deps/test-fd9873c3349180f5)
==751552==    by 0x13CF3A: core::ops::function::FnOnce::call_once (function.rs:227)
==751552==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==751552==

running 1 test
test create_test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

==751552==
==751552== HEAP SUMMARY:
==751552==     in use at exit: 0 bytes in 0 blocks
==751552==   total heap usage: 733 allocs, 733 frees, 499,671 bytes allocated
==751552==
==751552== All heap blocks were freed -- no leaks are possible
==751552==
==751552== For lists of detected and suppressed errors, rerun with: -s
==751552== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
```

miquels commented:
However, I noticed that I had to fix the test because of the change in `bitflags` so that flags are now in a sub-namespace instead of at the root of the crate.

What would you like to do about it? I can see two solutions:
- try to bring the old at-the-root consts back, maybe just by duplicating them with `use`
- increment the version to `0.2.0`.

I think incrementing the version to 0.2.0 is appropriate, but it's your decision.

miquels committed:
update README

miquels committed:
fix test

cholcombe973 commented:
Your test looks clean! Thanks for doing that.
Yeah I think just incrementing the version to `0.2.0` is fine :)

cholcombe973 commented on Cargo.toml:
```diff
@@ -1,12 +1,12 @@
-bitflags = "~0.8"
-gdbm-sys = "~0.2"
+bitflags = "~1.2"
+gdbm-sys = { version = "~0.3", git = "https://github.com/miquels/gdbm-sys" }
```
Since we merged the change for gdbm-sys do you want to update this now? If not I can do it later after the merge.

miquels replied:
Oh yes, sure. I'll increment the version to 0.2.0 as well.

miquels committed:
depend on gdbm-sys 0.3, up version to 0.2.0

miquels commented:
One more thing. I made Gdbm `Send`, this has UB / security implications if that is wrong. I feel I need to elaborate on why I think that is  correct.

I checked the C source for global variables, did not find them. There is one thread-local variable, the errno, but that is always written and then read on the same thread. The gdbm 'handle' is a pointer to malloc'ed space, so even if that contains self-refential data, it's not going to be moved by the Rust code, it's what in Rust would be called 'pinned'. So I'm quite sure that this is correct.

miquels committed:
correctness update for 'datum'

In several places a 'datum' is constructed without checking the
length of the Rust string we're converting into a datum. So now
we check it, and if it's longer than i32::MAX we error out.

Similarly, we should also check the validity of a 'datum'
we get from the C size. Not only check if content.dptr.is_null,
but also content.dsize < 0 and return an error.

Because of this we have to change the signature of the
'store' method to Result<bool, GdbmError> instead of -1/0/1.
But that is correct anyway, as -1 indicates an error.

Update test test  as well.

miquels committed:
replace method fdesc() with AsRawFd trait

We're bumping the version of this crate anyway, and
as_raw_fd() is more idiomatic than fdesc().

miquels committed:
Remove `Store` flags.

This was incorrect, there are only two store flags: INSERT and
REPLACE. The other constants were for options, which is not
implemented right now. So remove it and just pass a bool to store().

miquels committed:
use 'replace' instead of 'overwrite'

so that it is more in line with the gdbm constants.

miquels commented:
Hey, just checking in too see if there still is something to be done before you can pull this?

cholcombe973 reviewed (approved):
Great work!
